python/pyspark/worker.py

```diff
@@ -481,7 +482,12 @@ def main(infile, outfile):
 
         def process():
             iterator = deserializer.load_stream(infile)
-            serializer.dump_stream(func(split_index, iterator), outfile)
+            out_iter = func(split_index, iterator)
+            try:
+                serializer.dump_stream(out_iter, outfile)
+            finally:
+                if isinstance(out_iter, GeneratorType):
```

Comment:
@WeichenXu123 I thought out_iter is always a generater. Is there other cases? It would be great if we can add a simple test for mapPartitions.

Reply:
ah, okay, the func actually can return an arbitrary iterator that doesn't have `close`. We can even do `hasattr(out_iter, "close")` and call it in case it has a generator-like instance.